The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8170767174/job/22337571133) [(plain)](https://github.com/rust-lang/rust/commit/300c184d2cf0139da2e54e6da4baa5432dd51f48/checks/22337571133/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=chenyukang
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_8658edcd-04f0-4dd7-ac8b-cd7c0fa8b1c2
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=yukang-fix-121915-import
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_8658edcd-04f0-4dd7-ac8b-cd7c0fa8b1c2
GITHUB_REF=refs/pull/121958/merge
GITHUB_REF_NAME=121958/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=cf404bfe2cea38abff30a1e9a0dc00edc8be0066
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_8658edcd-04f0-4dd7-ac8b-cd7c0fa8b1c2
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_8658edcd-04f0-4dd7-ac8b-cd7c0fa8b1c2
GITHUB_TRIGGERING_ACTOR=chenyukang
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/121958/merge
GITHUB_WORKFLOW_SHA=cf404bfe2cea38abff30a1e9a0dc00edc8be0066
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#12 writing image sha256:dbd8f09ed0d046255d7f4d6fab3742c41b19efb5785867612cb3d0d7a6c37e6a done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Mar  6 10:45:21 UTC 2024
  network time: Wed, 06 Mar 2024 10:45:21 GMT
  network time: Wed, 06 Mar 2024 10:45:21 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
....................................................................i.................i.  8360/16267
.......................................................................................i  8448/16267
........................................................................................  8536/16267
....................................................i...................................  8624/16267
................................................i..................................F....  8712/16267
.....................................................FF.......F..FFF....................  8800/16267
.......................ii..............................................................i  8976/16267
i..................................................................i....................  9064/16267
........................................................................................  9152/16267
........................................................................................  9240/16267
---
+    |         ^^^^^^^^^^^^^^^^^^^^^
21 
- error: unused import: `TryInto`
-   --> $DIR/suggest-remove-issue-121315.rs:11:20
+ error: unused imports: `TryFrom`, `TryInto`
24    |
- LL | use std::convert::{TryInto};
-    |                    ^^^^^^^
+ LL |     use std::convert::{TryFrom, TryInto};
---

3 #![deny(unused_imports)]
4 #![allow(dead_code)]
5 
- // Test remove FlatUnused
- //~^ ERROR the item `TryFrom` is imported redundantly
+     // Test remove FlatUnused
+     
+     
+     //~^ ERROR unused import: `std::convert::TryFrom`
+     //~| ERROR the item `TryFrom` is imported redundantly
+     let _ = u32::try_from(5i32);
8 
8 
- // Test remove NestedFullUnused
- //~^ ERROR the item `TryInto` is imported redundantly
+ fn test1() {
+     // Test remove NestedFullUnused
+     
+     //~^ ERROR unused imports: `TryFrom`, `TryInto`
+     //~^^ ERROR the item `TryFrom` is imported redundantly
+     //~| ERROR the item `TryInto` is imported redundantly
- // Test remove NestedPartialUnused
- use std::convert::{Infallible};
- use std::convert::{Infallible};
- //~^ ERROR the item `From` is imported redundantly
+     let _ = u32::try_from(5i32);
+     let _a: i32 = u32::try_into(5u32).unwrap();
15 
15 
- // Test remove both redundant and unused?
- // use std::convert::{AsMut, Into};
+     // Test remove both redundant and unused
+     
+     
+     //~^ ERROR unused imports: `AsMut`, `Into`
+     //~| the item `Into` is imported redundantly
18 
- trait MyTrait {}
- impl MyTrait for fn() -> Infallible {}
+     let _a: u32 = (5u8).into();
21 
- fn main() {
-     let _ = u32::try_from(5i32);
-     let _ = u32::try_from(5i32);
-     let _a: i32 = u32::try_into(5u32).unwrap();
-     let _a: String = String::from("hello anan");
-     //let _a: u32 = (5i32).into();
+     // Test remove NestedPartialUnused
+     use std::convert::{Infallible};
+     use std::convert::{Infallible};
+     //~^ ERROR unused import: `From`
+ 
+     trait MyTrait {}
+     impl MyTrait for fn() -> Infallible {}
+ 
+ fn main() {}
28 



The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/suggest-remove-issue-121315/suggest-remove-issue-121315.fixed
To only update this specific test, also pass `--test-args imports/suggest-remove-issue-121315.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/suggest-remove-issue-121315.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/suggest-remove-issue-121315" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/suggest-remove-issue-121315/auxiliary" "--edition" "2021"
--- stderr -------------------------------
error: the item `TryFrom` is imported redundantly
##[error]  --> /checkout/tests/ui/imports/suggest-remove-issue-121315.rs:8:9
   |
---
   |
LL |     use std::convert::TryFrom;
   |         ^^^^^^^^^^^^^^^^^^^^^

error: unused imports: `TryFrom`, `TryInto`
   |
LL |     use std::convert::{TryFrom, TryInto};
   |                        ^^^^^^^  ^^^^^^^

---


---- [ui] tests/ui/imports/redundant-import-issue-121915-2015.rs stdout ----

error: /checkout/tests/ui/imports/redundant-import-issue-121915-2015.rs:8: unexpected error: '8:9: 8:25: unused import: `aux_issue_121915` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/redundant-import-issue-121915-2015.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/redundant-import-issue-121915-2015" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/redundant-import-issue-121915-2015/auxiliary" "--extern" "aux_issue_121915" "--edition" "2015"
    Error {
        line_num: 8,
        kind: Some(
            Error,
            Error,
        ),
        msg: "8:9: 8:25: unused import: `aux_issue_121915` [unused_imports]",
]

thread '[ui] tests/ui/imports/redundant-import-issue-121915-2015.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [ui] tests/ui/imports/redundant-import-issue-121915.rs stdout ----

error: /checkout/tests/ui/imports/redundant-import-issue-121915.rs:6: unexpected error: '6:9: 6:25: unused import: `aux_issue_121915` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/redundant-import-issue-121915.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/redundant-import-issue-121915" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/redundant-import-issue-121915/auxiliary" "--extern" "aux_issue_121915" "--edition" "2018"
    Error {
        line_num: 6,
        kind: Some(
            Error,
            Error,
        ),
        msg: "6:9: 6:25: unused import: `aux_issue_121915` [unused_imports]",
]

thread '[ui] tests/ui/imports/redundant-import-issue-121915.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/unused/issue-59896.rs stdout ----

error: /checkout/tests/ui/lint/unused/issue-59896.rs:6: unexpected error: '6:9: 6:10: unused import: `S` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/unused/issue-59896.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused/issue-59896" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused/issue-59896/auxiliary"
    Error {
        line_num: 6,
        kind: Some(
            Error,
            Error,
        ),
        msg: "6:9: 6:10: unused import: `S` [unused_imports]",
]

thread '[ui] tests/ui/lint/unused/issue-59896.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/use-redundant/use-redundant-glob.rs stdout ----

error: /checkout/tests/ui/lint/use-redundant/use-redundant-glob.rs:11: unexpected warning: '11:9: 11:17: unused import: `bar::Foo` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/use-redundant/use-redundant-glob.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-glob" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-glob/auxiliary"
unexpected errors (from JSON output): [
unexpected errors (from JSON output): [
    Error {
        line_num: 11,
        kind: Some(
            Warning,
        ),
        msg: "11:9: 11:17: unused import: `bar::Foo` [unused_imports]",
]

thread '[ui] tests/ui/lint/use-redundant/use-redundant-glob.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/use-redundant/use-redundant-glob-parent.rs stdout ----

error: /checkout/tests/ui/lint/use-redundant/use-redundant-glob-parent.rs:12: unexpected warning: '12:9: 12:17: unused import: `bar::Foo` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/use-redundant/use-redundant-glob-parent.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-glob-parent" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-glob-parent/auxiliary"
unexpected errors (from JSON output): [
unexpected errors (from JSON output): [
    Error {
        line_num: 12,
        kind: Some(
            Warning,
        ),
        msg: "12:9: 12:17: unused import: `bar::Foo` [unused_imports]",
]

thread '[ui] tests/ui/lint/use-redundant/use-redundant-glob-parent.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/use-redundant/use-redundant-issue-71450.rs stdout ----

error: /checkout/tests/ui/lint/use-redundant/use-redundant-issue-71450.rs:26: unexpected warning: '26:13: 26:32: unused import: `std::string::String` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 0
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/use-redundant/use-redundant-issue-71450.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-issue-71450" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-issue-71450/auxiliary"
    Error {
        line_num: 26,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "26:13: 26:32: unused import: `std::string::String` [unused_imports]",
]

thread '[ui] tests/ui/lint/use-redundant/use-redundant-issue-71450.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/use-redundant/use-redundant-prelude-rust-2021.rs stdout ----

error: /checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2021.rs:5: unexpected warning: '5:5: 5:26: unused import: `std::convert::TryFrom` [unused_imports]'

error: /checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2021.rs:6: unexpected warning: '6:5: 6:26: unused import: `std::convert::TryInto` [unused_imports]'
error: 2 unexpected errors found, 0 expected errors not found
status: exit status: 0
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2021.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-prelude-rust-2021" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-prelude-rust-2021/auxiliary" "--edition=2021"
    Error {
        line_num: 5,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "5:5: 5:26: unused import: `std::convert::TryFrom` [unused_imports]",
    Error {
        line_num: 6,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "6:5: 6:26: unused import: `std::convert::TryInto` [unused_imports]",
]

thread '[ui] tests/ui/lint/use-redundant/use-redundant-prelude-rust-2021.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/use-redundant/use-redundant.rs stdout ----

error: /checkout/tests/ui/lint/use-redundant/use-redundant.rs:21: unexpected warning: '21:9: 21:24: unused import: `crate::foo::Bar` [unused_imports]'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/use-redundant/use-redundant.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant/auxiliary"
unexpected errors (from JSON output): [
unexpected errors (from JSON output): [
    Error {
        line_num: 21,
        kind: Some(
            Warning,
        ),
        msg: "21:9: 21:24: unused import: `crate::foo::Bar` [unused_imports]",
]

thread '[ui] tests/ui/lint/use-redundant/use-redundant.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
explicit panic

---- [ui] tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs stdout ----

error: /checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs:5: unexpected warning: '5:5: 5:30: unused import: `std::option::Option::Some` [unused_imports]'

error: /checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs:6: unexpected warning: '6:5: 6:30: unused import: `std::option::Option::None` [unused_imports]'

error: /checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs:8: unexpected warning: '8:5: 8:28: unused import: `std::result::Result::Ok` [unused_imports]'

error: /checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs:9: unexpected warning: '9:5: 9:29: unused import: `std::result::Result::Err` [unused_imports]'
error: 4 unexpected errors found, 0 expected errors not found
status: exit status: 0
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-prelude-rust-2015" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/use-redundant/use-redundant-prelude-rust-2015/auxiliary"
    Error {
        line_num: 5,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "5:5: 5:30: unused import: `std::option::Option::Some` [unused_imports]",
    Error {
        line_num: 6,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "6:5: 6:30: unused import: `std::option::Option::None` [unused_imports]",
    Error {
        line_num: 8,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "8:5: 8:28: unused import: `std::result::Result::Ok` [unused_imports]",
    Error {
        line_num: 9,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "9:5: 9:29: unused import: `std::result::Result::Err` [unused_imports]",
]

thread '[ui] tests/ui/lint/use-redundant/use-redundant-prelude-rust-2015.rs' panicked at src/tools/compiletest/src/runtest.rs:1847:13:
explicit panic
```

</details>

--- a/compiler/rustc_lint/src/context/diagnostics.rs
+++ b/compiler/rustc_lint/src/context/diagnostics.rs
@@ -143,7 +143,11 @@ pub(super) fn builtin(sess: &Session, diagnostic: BuiltinLintDiag, diag: &mut Di
         BuiltinLintDiag::RedundantImport(spans, ident) => {
             for (span, is_imported) in spans {
                 let introduced = if is_imported { "imported" } else { "defined" };
-                diag.span_label(span, format!("the item `{ident}` is already {introduced} here"));
+                let span_msg = if span.is_dummy() { "by prelude" } else { "here" };
+                diag.span_label(
+                    span,
+                    format!("the item `{ident}` is already {introduced} {span_msg}"),
+                );
             }
         }
         BuiltinLintDiag::DeprecatedMacro(suggestion, span) => {

--- a/compiler/rustc_resolve/src/check_unused.rs
+++ b/compiler/rustc_resolve/src/check_unused.rs
@@ -137,6 +137,81 @@ impl<'a, 'b, 'tcx> UnusedImportCheckVisitor<'a, 'b, 'tcx> {
             self.check_import_as_underscore(item, *id);
         }
     }
+
+    fn report_unused_extern_crate_items(
+        &mut self,
+        maybe_unused_extern_crates: FxHashMap<ast::NodeId, Span>,
+    ) {
+        let tcx = self.r.tcx();
+        for extern_crate in &self.extern_crate_items {
+            let warn_if_unused = !extern_crate.ident.name.as_str().starts_with('_');
+
+            // If the crate is fully unused, we suggest removing it altogether.
+            // We do this in any edition.
+            if warn_if_unused {
+                if let Some(&span) = maybe_unused_extern_crates.get(&extern_crate.id) {
+                    self.r.lint_buffer.buffer_lint_with_diagnostic(
+                        UNUSED_EXTERN_CRATES,
+                        extern_crate.id,
+                        span,
+                        "unused extern crate",
+                        BuiltinLintDiag::UnusedExternCrate {
+                            removal_span: extern_crate.span_with_attributes,
+                        },
+                    );
+                    continue;
+                }
+            }
+
+            // If we are not in Rust 2018 edition, then we don't make any further
+            // suggestions.
+            if !tcx.sess.at_least_rust_2018() {
+                continue;
+            }
+
+            // If the extern crate has any attributes, they may have funky
+            // semantics we can't faithfully represent using `use` (most
+            // notably `#[macro_use]`). Ignore it.
+            if extern_crate.has_attrs {
+                continue;
+            }
+
+            // If the extern crate is renamed, then we cannot suggest replacing it with a use as this
+            // would not insert the new name into the prelude, where other imports in the crate may be
+            // expecting it.
+            if extern_crate.renames {
+                continue;
+            }
+
+            // If the extern crate isn't in the extern prelude,
+            // there is no way it can be written as a `use`.
+            if !self
+                .r
+                .extern_prelude
+                .get(&extern_crate.ident)
+                .is_some_and(|entry| !entry.introduced_by_item)
+            {
+                continue;
+            }
+
+            let vis_span = extern_crate
+                .vis_span
+                .find_ancestor_inside(extern_crate.span)
+                .unwrap_or(extern_crate.vis_span);
+            let ident_span = extern_crate
+                .ident
+                .span
+                .find_ancestor_inside(extern_crate.span)
+                .unwrap_or(extern_crate.ident.span);
+            self.r.lint_buffer.buffer_lint_with_diagnostic(
+                UNUSED_EXTERN_CRATES,
+                extern_crate.id,
+                extern_crate.span,
+                "`extern crate` is not idiomatic in the new edition",
+                BuiltinLintDiag::ExternCrateNotIdiomatic { vis_span, ident_span },
+            );
+        }
+    }
 }
 
 impl<'a, 'b, 'tcx> Visitor<'a> for UnusedImportCheckVisitor<'a, 'b, 'tcx> {
@@ -335,6 +410,8 @@ impl Resolver<'_, '_> {
         };
         visit::walk_crate(&mut visitor, krate);
 
+        visitor.report_unused_extern_crate_items(maybe_unused_extern_crates);
+
         for unused in visitor.unused_imports.values() {
             let mut fixes = Vec::new();
             let spans = match calc_unused_spans(unused, &unused.use_tree, unused.use_tree_id) {
@@ -413,75 +490,6 @@ impl Resolver<'_, '_> {
                 ms,
                 msg,
                 BuiltinLintDiag::UnusedImports(fix_msg.into(), fixes, test_module_span),
-            );
-        }
-
-        for extern_crate in visitor.extern_crate_items {
-            let warn_if_unused = !extern_crate.ident.name.as_str().starts_with('_');
-
-            // If the crate is fully unused, we suggest removing it altogether.
-            // We do this in any edition.
-            if warn_if_unused {
-                if let Some(&span) = maybe_unused_extern_crates.get(&extern_crate.id) {
-                    visitor.r.lint_buffer.buffer_lint_with_diagnostic(
-                        UNUSED_EXTERN_CRATES,
-                        extern_crate.id,
-                        span,
-                        "unused extern crate",
-                        BuiltinLintDiag::UnusedExternCrate {
-                            removal_span: extern_crate.span_with_attributes,
-                        },
-                    );
-                    continue;
-                }
-            }
-
-            // If we are not in Rust 2018 edition, then we don't make any further
-            // suggestions.
-            if !tcx.sess.at_least_rust_2018() {
-                continue;
-            }
-
-            // If the extern crate has any attributes, they may have funky
-            // semantics we can't faithfully represent using `use` (most
-            // notably `#[macro_use]`). Ignore it.
-            if extern_crate.has_attrs {
-                continue;
-            }
-
-            // If the extern crate is renamed, then we cannot suggest replacing it with a use as this
-            // would not insert the new name into the prelude, where other imports in the crate may be
-            // expecting it.
-            if extern_crate.renames {
-                continue;
-            }
-
-            // If the extern crate isn't in the extern prelude,
-            // there is no way it can be written as a `use`.
-            if !visitor
-                .r
-                .extern_prelude
-                .get(&extern_crate.ident)
-                .is_some_and(|entry| !entry.introduced_by_item)
-            {
-                continue;
-            }
-
-            let vis_span = extern_crate
-                .vis_span
-                .find_ancestor_inside(extern_crate.span)
-                .unwrap_or(extern_crate.vis_span);
-            let ident_span = extern_crate
-                .ident
-                .span
-                .find_ancestor_inside(extern_crate.span)
-                .unwrap_or(extern_crate.ident.span);
-            visitor.r.lint_buffer.buffer_lint_with_diagnostic(
-                UNUSED_EXTERN_CRATES,
-                extern_crate.id,
-                extern_crate.span,
-                "`extern crate` is not idiomatic in the new edition",
-                BuiltinLintDiag::ExternCrateNotIdiomatic { vis_span, ident_span },
             );
         }
 

--- a/compiler/rustc_resolve/src/imports.rs
+++ b/compiler/rustc_resolve/src/imports.rs
@@ -1336,9 +1336,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         }
 
         let mut is_redundant = true;
-
         let mut redundant_span = PerNS { value_ns: None, type_ns: None, macro_ns: None };
-
         self.per_ns(|this, ns| {
             if is_redundant && let Ok(binding) = source_bindings[ns].get() {
                 if binding.res() == Res::Err {

--- a/tests/ui/imports/auxiliary/aux-issue-121915.rs
+++ b/tests/ui/imports/auxiliary/aux-issue-121915.rs
@@ -1,0 +1,1 @@
+pub fn item() {}

--- a/tests/ui/imports/redundant-import-issue-121915-2015.rs
+++ b/tests/ui/imports/redundant-import-issue-121915-2015.rs
@@ -1,0 +1,11 @@
+//@ compile-flags: --extern aux_issue_121915 --edition 2015
+//@ aux-build: aux-issue-121915.rs
+
+extern crate aux_issue_121915;
+
+#[deny(unused_imports)]
+fn main() {
+    use aux_issue_121915;
+    //~^ ERROR the item `aux_issue_121915` is imported redundantly
+    aux_issue_121915::item();
+}

--- a/tests/ui/imports/redundant-import-issue-121915-2015.stderr
+++ b/tests/ui/imports/redundant-import-issue-121915-2015.stderr
@@ -1,0 +1,17 @@
+error: the item `aux_issue_121915` is imported redundantly
+  --> $DIR/redundant-import-issue-121915-2015.rs:8:9
+   |
+LL | extern crate aux_issue_121915;
+   | ------------------------------ the item `aux_issue_121915` is already imported here
+...
+LL |     use aux_issue_121915;
+   |         ^^^^^^^^^^^^^^^^
+   |
+note: the lint level is defined here
+  --> $DIR/redundant-import-issue-121915-2015.rs:6:8
+   |
+LL | #[deny(unused_imports)]
+   |        ^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/imports/redundant-import-issue-121915.rs
+++ b/tests/ui/imports/redundant-import-issue-121915.rs
@@ -1,0 +1,9 @@
+//@ compile-flags: --extern aux_issue_121915 --edition 2018
+//@ aux-build: aux-issue-121915.rs
+
+#[deny(unused_imports)]
+fn main() {
+    use aux_issue_121915;
+    //~^ ERROR the item `aux_issue_121915` is imported redundantly
+    aux_issue_121915::item();
+}

--- a/tests/ui/imports/redundant-import-issue-121915.stderr
+++ b/tests/ui/imports/redundant-import-issue-121915.stderr
@@ -1,0 +1,14 @@
+error: the item `aux_issue_121915` is imported redundantly
+  --> $DIR/redundant-import-issue-121915.rs:6:9
+   |
+LL |     use aux_issue_121915;
+   |         ^^^^^^^^^^^^^^^^ the item `aux_issue_121915` is already defined by prelude
+   |
+note: the lint level is defined here
+  --> $DIR/redundant-import-issue-121915.rs:4:8
+   |
+LL | #[deny(unused_imports)]
+   |        ^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/imports/suggest-remove-issue-121315.rs
+++ b/tests/ui/imports/suggest-remove-issue-121315.rs
@@ -1,0 +1,40 @@
+//@ compile-flags: --edition 2021
+#![deny(unused_imports)]
+#![allow(dead_code)]
+
+fn test0() {
+    // Test remove FlatUnused
+    use std::convert::TryFrom;
+    //~^ ERROR the item `TryFrom` is imported redundantly
+    let _ = u32::try_from(5i32);
+}
+
+fn test1() {
+    // FIXME(yukang) Test remove NestedFullUnused
+    use std::convert::{TryFrom, TryInto};
+    //~^ ERROR the item `TryFrom` is imported redundantly
+    //~| ERROR the item `TryInto` is imported redundantly
+
+    let _ = u32::try_from(5i32);
+    let _a: i32 = u32::try_into(5u32).unwrap();
+}
+
+fn test2() {
+    // FIXME(yukang): Test remove both redundant and unused
+    use std::convert::{AsMut, Into};
+    //~^ ERROR unused import: `AsMut`
+    //~| ERROR the item `Into` is imported redundantly
+
+    let _a: u32 = (5u8).into();
+}
+
+fn test3() {
+    // Test remove NestedPartialUnused
+    use std::convert::{From, Infallible};
+    //~^ ERROR unused import: `From`
+
+    trait MyTrait {}
+    impl MyTrait for fn() -> Infallible {}
+}
+
+fn main() {}

--- a/tests/ui/imports/suggest-remove-issue-121315.stderr
+++ b/tests/ui/imports/suggest-remove-issue-121315.stderr
@@ -1,0 +1,56 @@
+error: the item `TryFrom` is imported redundantly
+  --> $DIR/suggest-remove-issue-121315.rs:7:9
+   |
+LL |     use std::convert::TryFrom;
+   |         ^^^^^^^^^^^^^^^^^^^^^
+  --> $SRC_DIR/std/src/prelude/mod.rs:LL:COL
+   |
+   = note: the item `TryFrom` is already defined here
+   |
+note: the lint level is defined here
+  --> $DIR/suggest-remove-issue-121315.rs:2:9
+   |
+LL | #![deny(unused_imports)]
+   |         ^^^^^^^^^^^^^^
+
+error: the item `TryFrom` is imported redundantly
+  --> $DIR/suggest-remove-issue-121315.rs:14:24
+   |
+LL |     use std::convert::{TryFrom, TryInto};
+   |                        ^^^^^^^
+  --> $SRC_DIR/std/src/prelude/mod.rs:LL:COL
+   |
+   = note: the item `TryFrom` is already defined here
+
+error: the item `TryInto` is imported redundantly
+  --> $DIR/suggest-remove-issue-121315.rs:14:33
+   |
+LL |     use std::convert::{TryFrom, TryInto};
+   |                                 ^^^^^^^
+  --> $SRC_DIR/std/src/prelude/mod.rs:LL:COL
+   |
+   = note: the item `TryInto` is already defined here
+
+error: unused import: `AsMut`
+  --> $DIR/suggest-remove-issue-121315.rs:24:24
+   |
+LL |     use std::convert::{AsMut, Into};
+   |                        ^^^^^
+
+error: the item `Into` is imported redundantly
+  --> $DIR/suggest-remove-issue-121315.rs:24:31
+   |
+LL |     use std::convert::{AsMut, Into};
+   |                               ^^^^
+  --> $SRC_DIR/std/src/prelude/mod.rs:LL:COL
+   |
+   = note: the item `Into` is already defined here
+
+error: unused import: `From`
+  --> $DIR/suggest-remove-issue-121315.rs:33:24
+   |
+LL |     use std::convert::{From, Infallible};
+   |                        ^^^^
+
+error: aborting due to 6 previous errors
+

